calico_containers/pycalico/block.py

```diff
@@ -0,0 +1,494 @@
+        be JSON serializable.
+        :return: None.
+        """
+        assert isinstance(address, IPAddress)
+        # Convert to an ordinal
+        ordinal = int(address - self.cidr.first)
+        assert 0 <= ordinal <= BLOCK_SIZE, "Address not in block."
+
+        # Check if allocated
+        if self.allocations[ordinal] is not None:
+            raise AlreadyAssignedError("%s is already assigned in block %s" % (
+                address, self.cidr))
+
+        # Set up attributes
+        attr_index = self._find_or_add_attrs(handle_id, attributes)
+        self.allocations[ordinal] = attr_index
+        return
+
+    def count_free_addresses(self):
+        """
+        Count the number of free addresses in this block.
+        :return: Number of free addresses.
+        """
+        count = 0
+        for a in self.allocations:
+            if a is None:
+                count += 1
+        return count
+
+    def release(self, addresses):
+        """
+        Release the given addresses.
+
+        :param addresses: Set of IPAddresses to release.
+        :return: (unallocated, handles_with_counts) Where:
+          - unallocted is a set of IPAddresses.  If any of the requested
+            addresses were not allocated, they are returned so the caller can
+            handle appropriately.
+          - handles_with_counts is a dictionary of handle_ids and the number of
+            addresses released for that handle.  They are returned so the
+            caller can decrement the affected handles.
+        """
+        assert isinstance(addresses, (set, frozenset))
+        deleting_ref_counts = {}
+        ordinals = []
+        unallocated = set()
+        handles_with_counts = {}
+        for address in addresses:
+            assert isinstance(address, IPAddress)
+            # Convert to an ordinal
+            ordinal = int(address - self.cidr.first)
+            assert 0 <= ordinal <= BLOCK_SIZE, "Address not in block."
+
+            # Check if allocated
+            attr_idx = self.allocations[ordinal]
+            if attr_idx is None:
+                _log.warning("Asked to release %s in block %s, but it was not "
+                             "allocated.", address, self.cidr)
+                unallocated.add(address)
+                continue
+            ordinals.append(ordinal)
+            old_count = deleting_ref_counts.get(attr_idx, 0)
+            deleting_ref_counts[attr_idx] = old_count + 1
+
+            # Increment our count of addresses by handle.
+            handle_id = self.\
+                attributes[attr_idx][AllocationBlock.ATTR_HANDLE_ID]
+            handle_count = handles_with_counts.setdefault(handle_id, 0)
+            handle_count += 1
+            handles_with_counts[handle_id] = handle_count
+
+        # Compute which attributes need to be cleaned up.  We do this by
+        # reference counting.  If we're deleting all the references, then it
+        # needs to be cleaned up.
```

Comment:
Can you add a quick comment to describe whether race conditions in the cleanup step are an issue here please? Did a bit of digging but it's not immediately clear how you handle the case where another process added a new reference after this process fetched its state from etcd. I think that's probably worth documenting.